File: usesense_flutter/ios/Classes/UseSenseFlutterPlugin.swift
import Flutter
import UIKit
import UseSenseSDK

/// Flutter plugin that bridges Dart calls to the UseSense iOS SDK.
///
/// Implements `UseSenseHostApi` (Pigeon-generated) for Dart → Native calls and
/// uses `UseSenseFlutterApiImpl` (Pigeon-generated) for Native → Dart callbacks.
public class UseSenseFlutterPlugin: NSObject, FlutterPlugin, UseSenseHostApi {

    private var flutterApi: UseSenseFlutterApiImpl?
    private var eventUnsubscribe: (() -> Void)?
    private var client: UseSenseClient?

    // MARK: - FlutterPlugin

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = UseSenseFlutterPlugin()
        instance.flutterApi = UseSenseFlutterApiImpl(binaryMessenger: registrar.messenger())
        UseSenseHostApiSetup(registrar.messenger(), instance)
    }

    // MARK: - UseSenseHostApi

    func initialize(config: PigeonUseSenseConfig, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            let environment: UseSenseEnvironment
            switch config.environment {
            case .sandbox:
                environment = .sandbox
            case .production:
                environment = .production
            case .auto:
                environment = .auto
            }

            var brandingConfig: BrandingConfig? = nil
            if let b = config.branding {
                brandingConfig = BrandingConfig(
                    displayName: b.displayName,
                    logoUrl: b.logoUrl,
                    primaryColor: b.primaryColor,
                    redirectUrl: b.redirectUrl,
                    buttonRadius: b.buttonRadius.map { Int($0) } ?? 12,
                    fontFamily: b.fontFamily
                )
            }

            let nativeConfig = UseSenseConfig(
                apiKey: config.apiKey,
                environment: environment,
                baseUrl: config.baseUrl,
                gatewayKey: config.gatewayKey,
                branding: brandingConfig
            )

            client = UseSense.initialize(config: nativeConfig)

            // Subscribe to native events and forward to Dart.
            eventUnsubscribe?.self()
            eventUnsubscribe = client?.onEvent { [weak self] event in
                DispatchQueue.main.async {
                    self?.forwardEvent(event)
                }
            }

            completion(.success(()))
        } catch {
            completion(.failure(PigeonError(
                code: "invalid_config",
                message: error.localizedDescription,
                details: nil
            )))
        }
    }

    func startVerification(request: PigeonVerificationRequest, completion: @escaping (Result<PigeonUseSenseResult, Error>) -> Void) {
        guard let client = client else {
            completion(.failure(PigeonError(
                code: "sdk_not_initialized",
                message: "UseSense SDK is not initialized. Call initialize() first.",
                details: nil
            )))
            return
        }

        guard let rootVC = UIApplication.shared.delegate?.window??.rootViewController else {
            completion(.failure(PigeonError(
                code: "sdk_not_initialized",
                message: "Root view controller is not available.",
                details: nil
            )))
            return
        }

        let sessionType: SessionType
        switch request.sessionType {
        case .enrollment:
            sessionType = .enrollment
        case .authentication:
            sessionType = .authentication
        }

        let nativeRequest = VerificationRequest(
            sessionType: sessionType,
            externalUserId: request.externalUserId,
            identityId: request.identityId,
            metadata: request.metadata?.mapValues { AnyCodableValue.string($0) }
        )

        let session = client.startVerification(request: nativeRequest)

        session.present(from: rootVC) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let useSenseResult):
                    let pigeonResult = PigeonUseSenseResult(
                        sessionId: useSenseResult.sessionId,
                        sessionType: useSenseResult.sessionType,
                        identityId: useSenseResult.identityId,
                        decision: useSenseResult.decision,
                        timestamp: useSenseResult.timestamp
                    )
                    completion(.success(pigeonResult))
                case .failure(let error):
                    if case .cancelled = error as? UseSenseSessionError {
                        self?.flutterApi?.onCancelled { _ in }
                        completion(.failure(PigeonError(
                            code: "session_cancelled",
                            message: "User cancelled the verification session.",
                            details: nil
                        )))
                    } else {
                        completion(.failure(self?.mapError(error) ?? PigeonError(
                            code: "sdk_error",
                            message: error.localizedDescription,
                            details: nil
                        )))
                    }
                }
            }
        }
    }

    func startRemoteEnrollment(remoteEnrollmentId: String, completion: @escaping (Result<PigeonUseSenseResult, Error>) -> Void) {
        guard let client = client else {
            completion(.failure(PigeonError(
                code: "sdk_not_initialized",
                message: "UseSense SDK is not initialized. Call initialize() first.",
                details: nil
            )))
            return
        }

        guard let rootVC = UIApplication.shared.delegate?.window??.rootViewController else {
            completion(.failure(PigeonError(
                code: "sdk_not_initialized",
                message: "Root view controller is not available.",
                details: nil
            )))
            return
        }

        client.startRemoteEnrollment(enrollmentId: remoteEnrollmentId, from: rootVC) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let useSenseResult):
                    completion(.success(PigeonUseSenseResult(
                        sessionId: useSenseResult.sessionId,
                        sessionType: useSenseResult.sessionType,
                        identityId: useSenseResult.identityId,
                        decision: useSenseResult.decision,
                        timestamp: useSenseResult.timestamp
                    )))
                case .failure(let error):
                    completion(.failure(self?.mapError(error) ?? PigeonError(
                        code: "sdk_error",
                        message: error.localizedDescription,
                        details: nil
                    )))
                }
            }
        }
    }

    func startRemoteVerification(remoteSessionId: String, completion: @escaping (Result<PigeonUseSenseResult, Error>) -> Void) {
        guard let client = client else {
            completion(.failure(PigeonError(
                code: "sdk_not_initialized",
                message: "UseSense SDK is not initialized. Call initialize() first.",
                details: nil
            )))
            return
        }

        guard let rootVC = UIApplication.shared.delegate?.window??.rootViewController else {
            completion(.failure(PigeonError(
                code: "sdk_not_initialized",
                message: "Root view controller is not available.",
                details: nil
            )))
            return
        }

        client.startRemoteVerification(sessionId: remoteSessionId, from: rootVC) { [weak self] result in
            DispatchQueue.main.async {
                switch result {
                case .success(let useSenseResult):
                    completion(.success(PigeonUseSenseResult(
                        sessionId: useSenseResult.sessionId,
                        sessionType: useSenseResult.sessionType,
                        identityId: useSenseResult.identityId,
                        decision: useSenseResult.decision,
                        timestamp: useSenseResult.timestamp
                    )))
                case .failure(let error):
                    completion(.failure(self?.mapError(error) ?? PigeonError(
                        code: "sdk_error",
                        message: error.localizedDescription,
                        details: nil
                    )))
                }
            }
        }
    }

    func isInitialized() throws -> Bool {
        return client != nil
    }

    func reset() throws {
        eventUnsubscribe?()
        eventUnsubscribe = nil
        client?.reset()
        client = nil
    }

    // MARK: - Private helpers

    private func forwardEvent(_ event: UseSenseEvent) {
        let pigeonType: PigeonEventType
        switch event.type {
        case .sessionCreated: pigeonType = .sessionCreated
        case .permissionsRequested: pigeonType = .permissionsRequested
        case .permissionsGranted: pigeonType = .permissionsGranted
        case .permissionsDenied: pigeonType = .permissionsDenied
        case .captureStarted: pigeonType = .captureStarted
        case .frameCaptured: pigeonType = .frameCaptured
        case .captureCompleted: pigeonType = .captureCompleted
        case .audioRecordStarted: pigeonType = .audioRecordStarted
        case .audioRecordCompleted: pigeonType = .audioRecordCompleted
        case .challengeStarted: pigeonType = .challengeStarted
        case .challengeCompleted: pigeonType = .challengeCompleted
        case .uploadStarted: pigeonType = .uploadStarted
        case .uploadProgress: pigeonType = .uploadProgress
        case .uploadCompleted: pigeonType = .uploadCompleted
        case .completeStarted: pigeonType = .completeStarted
        case .decisionReceived: pigeonType = .decisionReceived
        case .imageQualityCheck: pigeonType = .imageQualityCheck
        case .error: pigeonType = .error
        @unknown default: pigeonType = .error
        }

        // Convert data values to platform-channel-safe types.
        let safeData: [String: Any?]? = event.data?.mapValues { value in
            switch value {
            case let s as String: return s
            case let i as Int: return i
            case let d as Double: return d
            case let b as Bool: return b
            case nil: return nil
            default: return "\(value)"
            }
        }

        let pigeonEvent = PigeonUseSenseEvent(
            type: pigeonType,
            timestamp: Int64(event.timestamp),
            data: safeData
        )
        flutterApi?.onEvent(event: pigeonEvent) { _ in }
    }

    private func mapError(_ error: Error) -> PigeonError {
        if let sdkError = error as? UseSenseError {
            let code: String
            switch sdkError.code {
            case 1001: code = "camera_unavailable"
            case 1002: code = "camera_permission_denied"
            case 1003: code = "microphone_permission_denied"
            case 2001: code = "network_error"
            case 2002: code = "network_timeout"
            case 3001: code = "session_expired"
            case 3002: code = "upload_failed"
            case 4001: code = "capture_failed"
            case 4002: code = "encoding_failed"
            case 5001: code = "invalid_config"
            case 6001: code = "quota_exceeded"
            default: code = "sdk_error"
            }
            return PigeonError(code: code, message: sdkError.message, details: nil)
        }
        return PigeonError(code: "sdk_error", message: error.localizedDescription, details: nil)
    }
}

/// Enum matching the iOS SDK's session error type for cancellation detection.
enum UseSenseSessionError: Error {
    case cancelled
}

File: usesense_flutter/android/src/main/kotlin/com/usesense/flutter/UseSenseFlutterPlugin.kt
package com.usesense.flutter

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import com.usesense.sdk.UseSense
import com.usesense.sdk.UseSenseCallback
import com.usesense.sdk.UseSenseConfig
import com.usesense.sdk.UseSenseResult
import com.usesense.sdk.UseSenseError
import com.usesense.sdk.UseSenseEvent
import com.usesense.sdk.BrandingConfig
import com.usesense.sdk.EventType
import com.usesense.sdk.SessionType
import com.usesense.sdk.UseSenseEnvironment
import com.usesense.sdk.VerificationRequest

/**
 * Flutter plugin that bridges Dart calls to the UseSense Android SDK.
 *
 * Implements [UseSenseHostApi] (Pigeon-generated) for Dart → Native calls and
 * uses [UseSenseFlutterApi] (Pigeon-generated) for Native → Dart callbacks.
 */
class UseSenseFlutterPlugin : FlutterPlugin, ActivityAware, UseSenseHostApi {

    private var context: Context? = null
    private var activity: Activity? = null
    private var flutterApi: UseSenseFlutterApi? = null
    private var eventUnsubscribe: (() -> Unit)? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    // -------------------------------------------------------------------------
    // FlutterPlugin
    // -------------------------------------------------------------------------

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        flutterApi = UseSenseFlutterApi(binding.binaryMessenger)
        UseSenseHostApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        UseSenseHostApi.setUp(binding.binaryMessenger, null)
        eventUnsubscribe?.invoke()
        eventUnsubscribe = null
        flutterApi = null
        context = null
    }

    // -------------------------------------------------------------------------
    // ActivityAware
    // -------------------------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // -------------------------------------------------------------------------
    // UseSenseHostApi implementation
    // -------------------------------------------------------------------------

    override fun initialize(
        config: PigeonUseSenseConfig,
        callback: (Result<Unit>) -> Unit,
    ) {
        val ctx = context
        if (ctx == null) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "Android context is not available.", null)))
            return
        }
        try {
            val brandingConfig = config.branding?.let { b ->
                BrandingConfig(
                    displayName = b.displayName,
                    logoUrl = b.logoUrl,
                    primaryColor = b.primaryColor,
                    redirectUrl = b.redirectUrl,
                    buttonRadius = b.buttonRadius?.toInt() ?: 12,
                    fontFamily = b.fontFamily,
                )
            }
            val environment = when (config.environment) {
                PigeonUseSenseEnvironment.SANDBOX -> UseSenseEnvironment.SANDBOX
                PigeonUseSenseEnvironment.PRODUCTION -> UseSenseEnvironment.PRODUCTION
                PigeonUseSenseEnvironment.AUTO -> UseSenseEnvironment.AUTO
            }
            val nativeConfig = UseSenseConfig(
                apiKey = config.apiKey,
                environment = environment,
                baseUrl = config.baseUrl ?: UseSenseConfig.DEFAULT_BASE_URL,
                gatewayKey = config.gatewayKey,
                branding = brandingConfig,
                googleCloudProjectNumber = config.googleCloudProjectNumber
                    ?: UseSenseConfig.DEFAULT_GOOGLE_CLOUD_PROJECT_NUMBER,
            )
            UseSense.initialize(ctx, nativeConfig)

            // Subscribe to native events and forward them to Dart via FlutterApi.
            eventUnsubscribe?.invoke()
            eventUnsubscribe = UseSense.onEvent { event ->
                mainHandler.post {
                    flutterApi?.onEvent(mapEventToPigeon(event)) {}
                }
            }

            callback(Result.success(Unit))
        } catch (e: Exception) {
            callback(Result.failure(FlutterError("invalid_config", e.message, null)))
        }
    }

    override fun startVerification(
        request: PigeonVerificationRequest,
        callback: (Result<PigeonUseSenseResult>) -> Unit,
    ) {
        val currentActivity = activity
        if (currentActivity == null) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "Activity is not available. Ensure the plugin is attached to an Activity.", null)))
            return
        }
        if (!UseSense.isInitialized) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "UseSense SDK is not initialized. Call initialize() first.", null)))
            return
        }

        val sessionType = when (request.sessionType) {
            PigeonSessionType.ENROLLMENT -> SessionType.ENROLLMENT
            PigeonSessionType.AUTHENTICATION -> SessionType.AUTHENTICATION
        }

        val nativeRequest = VerificationRequest(
            sessionType = sessionType,
            externalUserId = request.externalUserId,
            identityId = request.identityId,
            metadata = request.metadata?.let { map ->
                @Suppress("UNCHECKED_CAST")
                map as Map<String, Any>
            },
        )

        UseSense.startVerification(currentActivity, nativeRequest, object : UseSenseCallback {
            override fun onSuccess(result: UseSenseResult) {
                mainHandler.post {
                    callback(Result.success(mapResultToPigeon(result)))
                }
            }

            override fun onError(error: UseSenseError) {
                mainHandler.post {
                    callback(Result.failure(mapErrorToFlutter(error)))
                }
            }

            override fun onCancelled() {
                mainHandler.post {
                    flutterApi?.onCancelled {}
                    callback(Result.failure(FlutterError("session_cancelled", "User cancelled the verification session.", null)))
                }
            }
        })
    }

    override fun startRemoteEnrollment(
        remoteEnrollmentId: String,
        callback: (Result<PigeonUseSenseResult>) -> Unit,
    ) {
        val ctx = context
        if (ctx == null) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "Android context is not available.", null)))
            return
        }
        if (!UseSense.isInitialized) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "UseSense SDK is not initialized. Call initialize() first.", null)))
            return
        }

        // The native SDK handles the hosted page flow internally.
        // We use startRemoteEnrollment which launches a HostedPageActivity.
        // The callback is handled via the event system – we listen for the
        // DECISION_RECEIVED event to resolve this future.
        try {
            UseSense.startRemoteEnrollment(ctx, remoteEnrollmentId)
            // Remote flows resolve via events – we set up a one-shot listener
            // to capture the result and resolve the callback.
            listenForSessionResult(callback)
        } catch (e: Exception) {
            callback(Result.failure(FlutterError("sdk_error", e.message, null)))
        }
    }

    override fun startRemoteVerification(
        remoteSessionId: String,
        callback: (Result<PigeonUseSenseResult>) -> Unit,
    ) {
        val ctx = context
        if (ctx == null) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "Android context is not available.", null)))
            return
        }
        if (!UseSense.isInitialized) {
            callback(Result.failure(FlutterError("sdk_not_initialized", "UseSense SDK is not initialized. Call initialize() first.", null)))
            return
        }

        try {
            UseSense.startRemoteVerification(ctx, remoteSessionId)
            listenForSessionResult(callback)
        } catch (e: Exception) {
            callback(Result.failure(FlutterError("sdk_error", e.message, null)))
        }
    }

    override fun isInitialized(): Boolean {
        return UseSense.isInitialized
    }

    override fun reset() {
        eventUnsubscribe?.invoke()
        eventUnsubscribe = null
        UseSense.reset()
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private fun listenForSessionResult(callback: (Result<PigeonUseSenseResult>) -> Unit) {
        var unsubscribe: (() -> Unit)? = null
        unsubscribe = UseSense.onEvent { event ->
            when (event.type) {
                EventType.DECISION_RECEIVED -> {
                    unsubscribe?.invoke()
                    val data = event.data ?: emptyMap()
                    val result = PigeonUseSenseResult(
                        sessionId = data["sessionId"]?.toString() ?: "",
                        sessionType = data["sessionType"]?.toString(),
                        identityId = data["identityId"]?.toString(),
                        decision = data["decision"]?.toString() ?: "REJECT",
                        timestamp = data["timestamp"]?.toString() ?: "",
                    )
                    mainHandler.post { callback(Result.success(result)) }
                }
                EventType.ERROR -> {
                    unsubscribe?.invoke()
                    val msg = event.data?.get("message")?.toString() ?: "Session failed."
                    mainHandler.post {
                        callback(Result.failure(FlutterError("session_error", msg, null)))
                    }
                }
                else -> { /* ignore other events */ }
            }
        }
    }

    private fun mapEventToPigeon(event: UseSenseEvent): PigeonUseSenseEvent {
        val pigeonType = when (event.type) {
            EventType.SESSION_CREATED -> PigeonEventType.SESSION_CREATED
            EventType.PERMISSIONS_REQUESTED -> PigeonEventType.PERMISSIONS_REQUESTED
            EventType.PERMISSIONS_GRANTED -> PigeonEventType.PERMISSIONS_GRANTED
            EventType.PERMISSIONS_DENIED -> PigeonEventType.PERMISSIONS_DENIED
            EventType.CAPTURE_STARTED -> PigeonEventType.CAPTURE_STARTED
            EventType.FRAME_CAPTURED -> PigeonEventType.FRAME_CAPTURED
            EventType.CAPTURE_COMPLETED -> PigeonEventType.CAPTURE_COMPLETED
            EventType.AUDIO_RECORD_STARTED -> PigeonEventType.AUDIO_RECORD_STARTED
            EventType.AUDIO_RECORD_COMPLETED -> PigeonEventType.AUDIO_RECORD_COMPLETED
            EventType.CHALLENGE_STARTED -> PigeonEventType.CHALLENGE_STARTED
            EventType.CHALLENGE_COMPLETED -> PigeonEventType.CHALLENGE_COMPLETED
            EventType.UPLOAD_STARTED -> PigeonEventType.UPLOAD_STARTED
            EventType.UPLOAD_PROGRESS -> PigeonEventType.UPLOAD_PROGRESS
            EventType.UPLOAD_COMPLETED -> PigeonEventType.UPLOAD_COMPLETED
            EventType.COMPLETE_STARTED -> PigeonEventType.COMPLETE_STARTED
            EventType.DECISION_RECEIVED -> PigeonEventType.DECISION_RECEIVED
            EventType.IMAGE_QUALITY_CHECK -> PigeonEventType.IMAGE_QUALITY_CHECK
            EventType.ERROR -> PigeonEventType.ERROR
        }

        // Convert data map values to platform-channel-compatible types.
        val safeData = event.data?.mapValues { (_, v) ->
            when (v) {
                is String, is Int, is Long, is Double, is Boolean -> v
                null -> null
                else -> v.toString()
            }
        }

        return PigeonUseSenseEvent(
            type = pigeonType,
            timestamp = event.timestamp,
            data = safeData,
        )
    }

    private fun mapResultToPigeon(result: UseSenseResult): PigeonUseSenseResult {
        return PigeonUseSenseResult(
            sessionId = result.sessionId,
            sessionType = result.sessionType,
            identityId = result.identityId,
            decision = result.decision,
            timestamp = result.timestamp,
        )
    }

    private fun mapErrorToFlutter(error: UseSenseError): FlutterError {
        val code = when (error.code) {
            UseSenseError.CAMERA_UNAVAILABLE -> "camera_unavailable"
            UseSenseError.CAMERA_PERMISSION_DENIED -> "camera_permission_denied"
            UseSenseError.MICROPHONE_PERMISSION_DENIED -> "microphone_permission_denied"
            UseSenseError.NETWORK_ERROR -> "network_error"
            UseSenseError.NETWORK_TIMEOUT -> "network_timeout"
            UseSenseError.SESSION_EXPIRED -> "session_expired"
            UseSenseError.UPLOAD_FAILED -> "upload_failed"
            UseSenseError.CAPTURE_FAILED -> "capture_failed"
            UseSenseError.ENCODING_FAILED -> "encoding_failed"
            UseSenseError.INVALID_CONFIG -> "invalid_config"
            UseSenseError.QUOTA_EXCEEDED -> "quota_exceeded"
            else -> "sdk_error"
        }
        return FlutterError(code, error.message, error.details)
    }
}
